Avoid allocating same string "UTF-8" repeatedly

brentfulgham opened the pull request:
#### d64e7efdb2d92203173d89ac6038004e8e49d771
<pre>
Avoid allocating same string &quot;UTF-8&quot; repeatedly
<a href="https://bugs.webkit.org/show_bug.cgi?id=282612">https://bugs.webkit.org/show_bug.cgi?id=282612</a>
&lt;<a href="https://rdar.apple.com/138947605">rdar://138947605</a>&gt;

Reviewed by NOBODY (OOPS!).

Based on the Blink change by Dave Tapuska &lt;dtapuska@chromium.org&gt;:
<a href="https://chromium.googlesource.com/chromium/src/+/8643e19a2b29eebed35d6fe4f14882ca470dc9fe">https://chromium.googlesource.com/chromium/src/+/8643e19a2b29eebed35d6fe4f14882ca470dc9fe</a>

* Source/WebCore/PAL/pal/text/TextEncoding.h:
(PAL::TextEncoding::name const):
* Source/WebCore/PAL/pal/text/TextEncodingDetector.h:
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
(PAL::newTextCodec):
* Source/WTF/wtf/text/AtomString.cpp:
* Source/WTF/wtf/text/AtomString.h:
(WTF::utf8Atom):
* Source/WTF/wtf/text/StringImpl.cpp:
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::utf8Atom):
* Source/WTF/wtf/text/WTFString.cpp:
* Source/WTF/wtf/text/WTFString.h:
(WTF::utf8String):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d64e7efdb2d92203173d89ac6038004e8e49d771

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75042 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54473 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79485 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26292 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77159 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63609 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2258 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58920 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17185 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78109 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49071 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64466 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39296 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46414 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21971 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24619 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68195 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22310 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80965 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74301 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2362 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1455 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67174 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2512 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64483 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66462 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10405 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8575 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96571 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2327 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5131 "") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21103 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2354 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3278 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2362 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -272,8 +273,12 @@ Ref<StringImpl> StringImpl::create(std::span<const UChar> characters)
     return createInternal(characters);
 }
 
+//static const char* utfStr = "UTF-8";
```
There is debugging code left over. This PR is in draft though so maybe that's expected.

brentfulgham replied:
Yes. I left that in so I could restore it easily. This isn't really ready for review (I'm not sure if it's even a progression yet).

I could use your advise on whether the `TextEncodingRegistry` should accept AtomString as an argument, and bypass the String->StringView->ASCIILiteral->AtomString path.

cdumez commented on Source/WebCore/PAL/pal/text/TextEncoding.cpp:
```diff
@@ -176,7 +184,7 @@ const TextEncoding& UTF16LittleEndianEncoding()
 const TextEncoding& UTF8Encoding()
 {
-    static NeverDestroyed<TextEncoding> globalUTF8Encoding("UTF-8"_s);
+    static NeverDestroyed<TextEncoding> globalUTF8Encoding(utf8Atom());
```
Why is this better? It was calling the `TextEncoding` constructing that was taking an ASCIILiteral. It shouldn't have constructed a `String`.

brentfulgham replied:
It was definitely constructing a `StringImpl`. But perhaps it was the first time AtomString had seem `UTF-8` in the process since launch?

cdumez replied:
Can I see the backtrace of the StringImpl construction? I'm curious where it's at. It's not obvious to me. `"UTF-8"_s` constructs an ASCIILiteral and TextEncoder stores an ASCIILiteral. There should be no need to construct a StringImpl at all?

cdumez commented on Source/WebCore/PAL/pal/text/TextEncoding.h:
```diff
@@ -61,7 +62,7 @@ class TextEncoding : public WTF::URLTextEncoding {
     bool isUTF7Encoding() const;
 
-    ASCIILiteral m_name;
+    AtomString m_name;
```
Seems like a regression? The previous code never did any allocations since it was using ASCIILiteral, now we have a heap allocated object instead.

brentfulgham replied:
Something's not right, I think. These paths were hitting the StringImpl constructor with the `"UTF-8"` string. I can back this out and grab a trace showing what that looked like. Maybe I misunderstood what was going on?

cdumez replied:
Yes, I'm curious where the StringImpl construction was from. ASCIILiteral is just a literal `const char*`, not String/StringImpl involved.

brentfulgham committed:
Avoid allocating same string "UTF-8" repeatedly
https://bugs.webkit.org/show_bug.cgi?id=282612
<rdar://138947605>

Reviewed by NOBODY (OOPS!).

Based on the Blink change by Dave Tapuska <dtapuska@chromium.org>:
https://chromium.googlesource.com/chromium/src/+/8643e19a2b29eebed35d6fe4f14882ca470dc9fe

* Source/WebCore/PAL/pal/text/TextEncoding.h:
(PAL::TextEncoding::name const):
* Source/WebCore/PAL/pal/text/TextEncodingDetector.h:
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
(PAL::newTextCodec):
* Source/WTF/wtf/text/AtomString.cpp:
* Source/WTF/wtf/text/AtomString.h:
(WTF::utf8Atom):
* Source/WTF/wtf/text/StringImpl.cpp:
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::utf8Atom):
* Source/WTF/wtf/text/WTFString.cpp:
* Source/WTF/wtf/text/WTFString.h:
(WTF::utf8String):
